{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e56b4acb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "89921efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = pyensembl.ensembl_grch38.genes_by_name(\"NUTM1\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aa90aa4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Transcript(transcript_id='ENST00000537011', transcript_name='NUTM1-203', gene_id='ENSG00000184507', biotype='protein_coding', contig='15', start=34343315, end=34357735, strand='+', genome='GRCh38'),\n",
       " Transcript(transcript_id='ENST00000614490', transcript_name='NUTM1-204', gene_id='ENSG00000184507', biotype='protein_coding', contig='15', start=34343315, end=34357737, strand='+', genome='GRCh38'),\n",
       " Transcript(transcript_id='ENST00000438749', transcript_name='NUTM1-202', gene_id='ENSG00000184507', biotype='protein_coding', contig='15', start=34343410, end=34357737, strand='+', genome='GRCh38'),\n",
       " Transcript(transcript_id='ENST00000333756', transcript_name='NUTM1-201', gene_id='ENSG00000184507', biotype='protein_coding', contig='15', start=34345865, end=34357732, strand='+', genome='GRCh38')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ea71832",
   "metadata": {},
   "outputs": [],
   "source": [
    "t = g.transcripts[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5f182155",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MVVTLGPGPDCLILEASRQPQLVPKPERMASDGASALPGPDMSMKPSAAPSPSPALPFLPPTSDPPDHPPREPPPQPIMPSVFSPDNPLMLSAFPSSLLVTGDGGPCLSGAGAGKVIVKVKTEGGSAEPSQTQNFILTQTALNSTAPGTPCGGLEGPAPPFVTASNVKTILPSKAVGVSQEGPPGLPPQPPPPVAQLVPIVPLEKAWPGPHGTTGEGGPVATLSKPSLGDRSKISKDVYENFRQWQRYKALARRHLSQSPDTEALSCFLIPVLRSLARLKPTMTLEEGLPLAVQEWEHTSNFDRMIFYEMAERFMEFEAEEMQIQNTQLMNGSQGLSPATPLKLDPLGPLASEVCQQPVYIPKKAASKTRAPRRRQRKAQRPPAPEAPKEIPPEAVKEYVDIMEWLVGTHLATGESDGKQEEEGQQQEEEGMYPDPGLLSYINELCSQKVFVSKVEAVIHPQFLADLLSPEKQRDPLALIEELEQEEGLTLAQLVQKRLMALEEEEDAEAPPSFSGAQLDSSPSGSVEDEDGDGRLRPSPGLQGAGGAACLGKVSSSGKRAREVHGGQEQALDSPRGMHRDGNTLPSPSSWDLQPELAAPQGTPGPLGVERRGSGKVINQVSLHQDGHLGGAGPPGHCLVADRTSEALPLCWQGGFQPESTPSLDAGLAELAPLQGQGLEKQVLGLQKGQQTGGRGVLPQGKEPLAVPWEGSSGAMWGDDRGTPMAQSYDQNPSPRAAGERDDVCLSPGVWLSSEMDAVGLELPVQIEEVIESFQVEKCVTEYQEGCQGLGSRGNISLGPGETLVPGDTESSVIPCGGTVAAAALEKRNYCSLPGPLRANSPPLRSKENQEQSCETVGHPSDLWAEGCFPLLESGDSTLGSSKETLPPTCQGNLLIMGTEDASSLPEASQEAGSRGNSFSPLLETIEPVNILDVKDDCGLQLRVSEDTCPLNVHSYDPQGEGRVDPDLSKPKNLAPLQESQESYTTGTPKATSSHQGLGSTLPRRGTRNAIVPRETSVSKTHRSADRAKGKEKKKKEAEEEDEELSNFAYLLASKLSLSPREHPLSPHHASGGQGSQRASHLLPAGAKGPSKLPYPVAKSGKRALAGGPAPTEKTPHSGAQLGVPREKPLALGVVRPSQPRKRRCDSFVTGRRKKRRRSQ'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t.protein_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b166c54a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Untitled.ipynb                   \u001b[31mrun-mhcflurry-wildtype-nutm1.sh\u001b[m\u001b[m*\r\n",
      "predictions-nutm1.csv\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "278edd5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd; df = pd.read_csv(\"predictions-nutm1.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "26388950",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>pos</th>\n",
       "      <th>peptide</th>\n",
       "      <th>n_flank</th>\n",
       "      <th>c_flank</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>affinity</th>\n",
       "      <th>best_allele</th>\n",
       "      <th>affinity_percentile</th>\n",
       "      <th>processing_score</th>\n",
       "      <th>presentation_score</th>\n",
       "      <th>presentation_percentile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>267</td>\n",
       "      <td>FLIPVLRSL</td>\n",
       "      <td>EALSC</td>\n",
       "      <td>ARLKP</td>\n",
       "      <td>genotype_01</td>\n",
       "      <td>14.878352</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>0.041625</td>\n",
       "      <td>0.919071</td>\n",
       "      <td>0.993569</td>\n",
       "      <td>0.000489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>1123</td>\n",
       "      <td>VPREKPLAL</td>\n",
       "      <td>GAQLG</td>\n",
       "      <td>GVVRP</td>\n",
       "      <td>genotype_05</td>\n",
       "      <td>29.134906</td>\n",
       "      <td>HLA-B*07:02</td>\n",
       "      <td>0.000625</td>\n",
       "      <td>0.944343</td>\n",
       "      <td>0.988814</td>\n",
       "      <td>0.001685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>432</td>\n",
       "      <td>YPDPGLLSY</td>\n",
       "      <td>EEEGM</td>\n",
       "      <td>INELC</td>\n",
       "      <td>genotype_09</td>\n",
       "      <td>22.875828</td>\n",
       "      <td>HLA-B*35:01</td>\n",
       "      <td>0.003250</td>\n",
       "      <td>0.859776</td>\n",
       "      <td>0.987821</td>\n",
       "      <td>0.002065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>1123</td>\n",
       "      <td>VPREKPLAL</td>\n",
       "      <td>GAQLG</td>\n",
       "      <td>GVVRP</td>\n",
       "      <td>genotype_06</td>\n",
       "      <td>37.009156</td>\n",
       "      <td>HLA-B*08:01</td>\n",
       "      <td>0.003250</td>\n",
       "      <td>0.944343</td>\n",
       "      <td>0.985921</td>\n",
       "      <td>0.003125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>267</td>\n",
       "      <td>FLIPVLRSL</td>\n",
       "      <td>EALSC</td>\n",
       "      <td>ARLKP</td>\n",
       "      <td>genotype_10</td>\n",
       "      <td>35.967311</td>\n",
       "      <td>HLA-C*03:04</td>\n",
       "      <td>0.042625</td>\n",
       "      <td>0.919071</td>\n",
       "      <td>0.984941</td>\n",
       "      <td>0.003723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>267</td>\n",
       "      <td>FLIPVLRSL</td>\n",
       "      <td>EALSC</td>\n",
       "      <td>ARLKP</td>\n",
       "      <td>genotype_13</td>\n",
       "      <td>49.980725</td>\n",
       "      <td>HLA-C*06:02</td>\n",
       "      <td>0.055250</td>\n",
       "      <td>0.919071</td>\n",
       "      <td>0.979372</td>\n",
       "      <td>0.008424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>91</td>\n",
       "      <td>SAFPSSLLV</td>\n",
       "      <td>NPLML</td>\n",
       "      <td>TGDGG</td>\n",
       "      <td>genotype_10</td>\n",
       "      <td>37.818598</td>\n",
       "      <td>HLA-C*03:04</td>\n",
       "      <td>0.055250</td>\n",
       "      <td>0.808287</td>\n",
       "      <td>0.976119</td>\n",
       "      <td>0.012092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>1047</td>\n",
       "      <td>FAYLLASKL</td>\n",
       "      <td>EELSN</td>\n",
       "      <td>SLSPR</td>\n",
       "      <td>genotype_10</td>\n",
       "      <td>35.075759</td>\n",
       "      <td>HLA-C*03:04</td>\n",
       "      <td>0.038625</td>\n",
       "      <td>0.777191</td>\n",
       "      <td>0.975049</td>\n",
       "      <td>0.013098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>431</td>\n",
       "      <td>MYPDPGLLSY</td>\n",
       "      <td>QEEEG</td>\n",
       "      <td>INELC</td>\n",
       "      <td>genotype_15</td>\n",
       "      <td>32.006945</td>\n",
       "      <td>HLA-C*07:02</td>\n",
       "      <td>0.004375</td>\n",
       "      <td>0.737713</td>\n",
       "      <td>0.973522</td>\n",
       "      <td>0.015380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>1132</td>\n",
       "      <td>GVVRPSQPR</td>\n",
       "      <td>KPLAL</td>\n",
       "      <td>KRRCD</td>\n",
       "      <td>genotype_00</td>\n",
       "      <td>75.380438</td>\n",
       "      <td>HLA-A*11:01</td>\n",
       "      <td>0.388500</td>\n",
       "      <td>0.926014</td>\n",
       "      <td>0.970302</td>\n",
       "      <td>0.018723</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sequence_name   pos     peptide n_flank c_flank  sample_name   affinity  \\\n",
       "0    sequence_0   267   FLIPVLRSL   EALSC   ARLKP  genotype_01  14.878352   \n",
       "1    sequence_0  1123   VPREKPLAL   GAQLG   GVVRP  genotype_05  29.134906   \n",
       "2    sequence_0   432   YPDPGLLSY   EEEGM   INELC  genotype_09  22.875828   \n",
       "3    sequence_0  1123   VPREKPLAL   GAQLG   GVVRP  genotype_06  37.009156   \n",
       "4    sequence_0   267   FLIPVLRSL   EALSC   ARLKP  genotype_10  35.967311   \n",
       "5    sequence_0   267   FLIPVLRSL   EALSC   ARLKP  genotype_13  49.980725   \n",
       "6    sequence_0    91   SAFPSSLLV   NPLML   TGDGG  genotype_10  37.818598   \n",
       "7    sequence_0  1047   FAYLLASKL   EELSN   SLSPR  genotype_10  35.075759   \n",
       "8    sequence_0   431  MYPDPGLLSY   QEEEG   INELC  genotype_15  32.006945   \n",
       "9    sequence_0  1132   GVVRPSQPR   KPLAL   KRRCD  genotype_00  75.380438   \n",
       "\n",
       "   best_allele  affinity_percentile  processing_score  presentation_score  \\\n",
       "0  HLA-A*02:01             0.041625          0.919071            0.993569   \n",
       "1  HLA-B*07:02             0.000625          0.944343            0.988814   \n",
       "2  HLA-B*35:01             0.003250          0.859776            0.987821   \n",
       "3  HLA-B*08:01             0.003250          0.944343            0.985921   \n",
       "4  HLA-C*03:04             0.042625          0.919071            0.984941   \n",
       "5  HLA-C*06:02             0.055250          0.919071            0.979372   \n",
       "6  HLA-C*03:04             0.055250          0.808287            0.976119   \n",
       "7  HLA-C*03:04             0.038625          0.777191            0.975049   \n",
       "8  HLA-C*07:02             0.004375          0.737713            0.973522   \n",
       "9  HLA-A*11:01             0.388500          0.926014            0.970302   \n",
       "\n",
       "   presentation_percentile  \n",
       "0                 0.000489  \n",
       "1                 0.001685  \n",
       "2                 0.002065  \n",
       "3                 0.003125  \n",
       "4                 0.003723  \n",
       "5                 0.008424  \n",
       "6                 0.012092  \n",
       "7                 0.013098  \n",
       "8                 0.015380  \n",
       "9                 0.018723  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b3be367e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>pos</th>\n",
       "      <th>peptide</th>\n",
       "      <th>n_flank</th>\n",
       "      <th>c_flank</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>affinity</th>\n",
       "      <th>best_allele</th>\n",
       "      <th>affinity_percentile</th>\n",
       "      <th>processing_score</th>\n",
       "      <th>presentation_score</th>\n",
       "      <th>presentation_percentile</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>267</td>\n",
       "      <td>FLIPVLRSL</td>\n",
       "      <td>EALSC</td>\n",
       "      <td>ARLKP</td>\n",
       "      <td>genotype_01</td>\n",
       "      <td>14.878352</td>\n",
       "      <td>HLA-A*02:01</td>\n",
       "      <td>0.041625</td>\n",
       "      <td>0.919071</td>\n",
       "      <td>0.993569</td>\n",
       "      <td>0.000489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>1123</td>\n",
       "      <td>VPREKPLAL</td>\n",
       "      <td>GAQLG</td>\n",
       "      <td>GVVRP</td>\n",
       "      <td>genotype_05</td>\n",
       "      <td>29.134906</td>\n",
       "      <td>HLA-B*07:02</td>\n",
       "      <td>0.000625</td>\n",
       "      <td>0.944343</td>\n",
       "      <td>0.988814</td>\n",
       "      <td>0.001685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>1123</td>\n",
       "      <td>VPREKPLAL</td>\n",
       "      <td>GAQLG</td>\n",
       "      <td>GVVRP</td>\n",
       "      <td>genotype_06</td>\n",
       "      <td>37.009156</td>\n",
       "      <td>HLA-B*08:01</td>\n",
       "      <td>0.003250</td>\n",
       "      <td>0.944343</td>\n",
       "      <td>0.985921</td>\n",
       "      <td>0.003125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>267</td>\n",
       "      <td>FLIPVLRSL</td>\n",
       "      <td>EALSC</td>\n",
       "      <td>ARLKP</td>\n",
       "      <td>genotype_10</td>\n",
       "      <td>35.967311</td>\n",
       "      <td>HLA-C*03:04</td>\n",
       "      <td>0.042625</td>\n",
       "      <td>0.919071</td>\n",
       "      <td>0.984941</td>\n",
       "      <td>0.003723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sequence_0</td>\n",
       "      <td>267</td>\n",
       "      <td>FLIPVLRSL</td>\n",
       "      <td>EALSC</td>\n",
       "      <td>ARLKP</td>\n",
       "      <td>genotype_13</td>\n",
       "      <td>49.980725</td>\n",
       "      <td>HLA-C*06:02</td>\n",
       "      <td>0.055250</td>\n",
       "      <td>0.919071</td>\n",
       "      <td>0.979372</td>\n",
       "      <td>0.008424</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sequence_name   pos    peptide n_flank c_flank  sample_name   affinity  \\\n",
       "0    sequence_0   267  FLIPVLRSL   EALSC   ARLKP  genotype_01  14.878352   \n",
       "1    sequence_0  1123  VPREKPLAL   GAQLG   GVVRP  genotype_05  29.134906   \n",
       "3    sequence_0  1123  VPREKPLAL   GAQLG   GVVRP  genotype_06  37.009156   \n",
       "4    sequence_0   267  FLIPVLRSL   EALSC   ARLKP  genotype_10  35.967311   \n",
       "5    sequence_0   267  FLIPVLRSL   EALSC   ARLKP  genotype_13  49.980725   \n",
       "\n",
       "   best_allele  affinity_percentile  processing_score  presentation_score  \\\n",
       "0  HLA-A*02:01             0.041625          0.919071            0.993569   \n",
       "1  HLA-B*07:02             0.000625          0.944343            0.988814   \n",
       "3  HLA-B*08:01             0.003250          0.944343            0.985921   \n",
       "4  HLA-C*03:04             0.042625          0.919071            0.984941   \n",
       "5  HLA-C*06:02             0.055250          0.919071            0.979372   \n",
       "\n",
       "   presentation_percentile  \n",
       "0                 0.000489  \n",
       "1                 0.001685  \n",
       "3                 0.003125  \n",
       "4                 0.003723  \n",
       "5                 0.008424  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[(df.processing_score > 0.9) & (df.presentation_score > 0.9) & (df.affinity< 50)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e3bc777d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(\"FLIPVLRSL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b985e265",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
